### PR TITLE
Uncertainty: Quantile Regression

### PR DESCRIPTION
**Pull request by HansikaPH**

### CODE REVIEW

File: neuralprophet/configure.py [hunk 1]
```diff
@@ -110,23 +111,44 @@ class Train:
...
         if self.epochs is not None:
             self.lambda_delay = int(self.reg_delay_pct * self.epochs)
-        if type(self.loss_func) == str:
-            if self.loss_func.lower() in ["huber", "smoothl1", "smoothl1loss"]:
-                self.loss_func = torch.nn.SmoothL1Loss()
-            elif self.loss_func.lower() in ["mae", "l1", "l1loss"]:
-                self.loss_func = torch.nn.L1Loss()
-            elif self.loss_func.lower() in ["mse", "mseloss", "l2", "l2loss"]:
-                self.loss_func = torch.nn.MSELoss()
+        if self.quantiles is None:
```
Comment (ourownstory) [inline neuralprophet/configure.py:138]: the loss function specification should still be valid in the pinball loss

File: neuralprophet/custom_loss_metrics.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        Args:
+            outputs (torch.tensor): outputs from the model of dims (batch, n_quantiles, n_forecasts)
+            target (torch.tensor): actual targets of dims (batch, n_forecasts)
+
+        Returns:
+            pinball loss (float)
+        """
+        losses = []
+        for i, quantile in enumerate(self.quantiles):
+            error = target - outputs[:, i, :]
```
Comment (ourownstory) [inline neuralprophet/custom_loss_metrics.py:27]: error should be defined by loss function

File: neuralprophet/forecaster.py [hunk 3]
```diff
@@ -220,9 +230,11 @@ def _init_model(self):
             n_lags=self.n_lags,
             num_hidden_layers=self.config_model.num_hidden_layers,
             d_hidden=self.config_model.d_hidden,
+            n_quantiles=self.config_train.n_quantiles,
```
Comment (ourownstory) [inline neuralprophet/forecaster.py:233]: can remove?

File: neuralprophet/custom_loss_metrics.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import torch
+
+
+class PinballLoss(torch.nn.Module):
```
Comment (ourownstory) [inline neuralprophet/custom_loss_metrics.py:4]: could inherit from _Loss?
#229

File: neuralprophet/forecaster.py [hunk 5]
```diff
@@ -981,67 +996,97 @@ def predict(self, df):
...
+                    forecast = predicted[:, forecast_lag - 1]
+                pad_before = self.n_lags + forecast_lag - 1
+                pad_after = self.n_forecasts - forecast_lag
+                yhat = np.concatenate(([None] * pad_before, forecast, [None] * pad_after))
+                if self.quantiles_enabled:
+                    name = "yhat{} {}%".format(i + 1, self.config_train.quantiles[j] * 100)
+                else:
+                    name = "yhat{}".format(i + 1)
+                df_forecast[name] = yhat
+                if j == 0:
```
Comment (ourownstory) [inline neuralprophet/forecaster.py:1014]: does this actually retrieve the median quantile?

**Review by ourownstory — COMMENTED** [review]
Great work!
last things we might want to change
- allow any loss in pinball loss
- example notebook
- sanity check with few datasets

**Comment by ourownstory** [comment]
@HansikaPH Try rebasing with master now - the new training settings might help

File: neuralprophet/metrics.py [hunk 6]
```diff
@@ -239,41 +237,32 @@ def __init__(self, name=None, specific_column=None):
...
-            if median_quantile_index is not None:
-                predicted = predicted[:, :, self.specific_column]
-            else:
-                predicted = predicted[:, self.specific_column]
-            # if len(predicted.shape) == 2:
-            #     predicted = predicted.unsqueeze(1)
-            target = target[:, self.specific_column]
-        avg_value = self._update_batch_value(predicted, target, median_quantile_index, **kwargs)
+            predicted = predicted[:, :, self.specific_column].unsqueeze(dim=-1)
+            target = target[:, self.specific_column].unsqueeze(dim=-1)
```
Comment (ourownstory) [inline neuralprophet/metrics.py:253]: expected target should be 3D

File: neuralprophet/metrics.py [hunk 7]
```diff
@@ -239,41 +237,32 @@ def __init__(self, name=None, specific_column=None):
...
         if self.specific_column is not None:
-            if median_quantile_index is not None:
-                predicted = predicted[:, :, self.specific_column]
-            else:
-                predicted = predicted[:, self.specific_column]
-            # if len(predicted.shape) == 2:
-            #     predicted = predicted.unsqueeze(1)
-            target = target[:, self.specific_column]
-        avg_value = self._update_batch_value(predicted, target, median_quantile_index, **kwargs)
+            predicted = predicted[:, :, self.specific_column].unsqueeze(dim=-1)
```
Comment (ourownstory) [inline neuralprophet/metrics.py:252]: unsqueeze should happen before/after calling metrics, but not in metrics

File: neuralprophet/metrics.py [hunk 8]
```diff
@@ -300,14 +289,13 @@ def __init__(self, specific_column=None, shift_scale=None):
...
-    def _update_batch_value(self, predicted, target, median_quantile_index=None, **kwargs):
+    def _update_batch_value(self, predicted, target, **kwargs):
         predicted = predicted.numpy()
         target = target.numpy()
         if self.shift_scale is not None:
             predicted = self.shift_scale[1] * predicted + self.shift_scale[0]
             target = self.shift_scale[1] * target + self.shift_scale[0]
-        if median_quantile_index is not None:
-            predicted = predicted[:, median_quantile_index, :]
+        predicted = predicted[:, 0, :]
```
Comment (ourownstory) [inline neuralprophet/metrics.py:298]: indexing should happen before calling metrics

File: neuralprophet/plot_forecast.py [hunk 9]
```diff
@@ -59,15 +59,13 @@ def plot(
 
     if highlight_forecast is None or line_per_origin:
         for i, name in enumerate(yhat_col_names):
-            if quantiles is None:
+            if "50.0%" in name:
```
Comment (ourownstory) [inline neuralprophet/plot_forecast.py:62]: might no longer hold

File: neuralprophet/plot_forecast.py [hunk 10]
```diff
@@ -82,20 +80,17 @@ def plot(
                 y = fcst["yhat{}".format(i + 1)].values[-(1 + i + steps_from_last)]
                 ax.plot(x, y, "bx")
         else:
-            if quantiles is not None:
-                # for quantile in quantiles:
-                ax.plot(ds, fcst["yhat{} {}%".format(highlight_forecast, 0.5 * 100)], ls="-", c="b")
-                ax.plot(ds, fcst["yhat{} {}%".format(highlight_forecast, 0.5 * 100)], "bx")
+            ax.plot(ds, fcst["yhat{} {}%".format(highlight_forecast, 0.5 * 100)], ls="-", c="b")
```
Comment (ourownstory) [inline neuralprophet/plot_forecast.py:83]: can skip name appendix for median

File: neuralprophet/plot_forecast.py [hunk 11]
```diff
@@ -82,20 +80,17 @@ def plot(
...
+            ax.plot(ds, fcst["yhat{} {}%".format(highlight_forecast, 0.5 * 100)], "bx")
+
+            if len(quantiles) != 1:
```
Comment (ourownstory) [inline neuralprophet/plot_forecast.py:86]: plot for all quantiles band wrt median

File: neuralprophet/time_net.py [hunk 12]
```diff
@@ -322,20 +325,23 @@ def _compute_quantile_forecasts_from_diffs(self, diffs):
...
-        out[:, (median_quantile_index + 1) :, :] = upper_quantile_diffs + diffs.detach().clone()[
-            :, median_quantile_index, :
-        ].unsqueeze(dim=1).repeat(1, n_upper_quantiles, 1)
-        out[:, :median_quantile_index, :] = lower_quantile_diffs + diffs.detach().clone()[
-            :, median_quantile_index, :
-        ].unsqueeze(dim=1).repeat(1, n_lower_quantiles, 1)
+        if self.n_quantiles != 1:
+            # generate the actual quantile forecasts from predicted differences
+            quantiles_divider_index = next(i for i, q in enumerate(self.quantiles) if q > 0.5)
+            n_upper_quantiles = diffs.shape[1] - quantiles_divider_index
```
Comment (ourownstory) [inline neuralprophet/time_net.py:325]: What if only upper or lower quantiles?

File: neuralprophet/time_net.py [hunk 13]
```diff
@@ -322,20 +325,23 @@ def _compute_quantile_forecasts_from_diffs(self, diffs):
...
+            n_lower_quantiles = quantiles_divider_index - 1
+
+            upper_quantile_diffs = diffs[:, quantiles_divider_index:, :]
+            lower_quantile_diffs = -(diffs[:, 1:quantiles_divider_index, :])
+            out = diffs.clone()
```
Comment (ourownstory) [inline neuralprophet/time_net.py:330]: out = zeros_like(diffs)

Comment (ourownstory) [reply neuralprophet/time_net.py:330]: out[median] = diffs[median]
out[upper] = diffs[median].detach() + diffs[upper]
out[lower] = diffs[median].detach() - diffs[lower]

File: neuralprophet/time_net.py [hunk 14]
```diff
@@ -392,11 +398,11 @@ def trend(self, t, quantile=None):
             trend = self.trend_k0 * torch.unsqueeze(t, dim=1)
         else:
             trend = self._piecewise_linear_trend(t)
-        if quantile is None:
-            return self.bias + trend
-        else:
+        if quantile is not None:
```
Comment (ourownstory) [inline neuralprophet/time_net.py:396]: can we skip this and do quantile-indexing in the function calling this (outside)?

File: neuralprophet/time_net.py [hunk 15]
```diff
@@ -412,11 +418,11 @@ def seasonality(self, features, name, quantile=None):
         seasonality = torch.sum(
             torch.unsqueeze(features, dim=1) * torch.unsqueeze(self.season_params[name], dim=1), dim=3
         )
-        if quantile is None:
-            return seasonality
-        else:
+        if quantile is not None:
```
Comment (ourownstory) [inline neuralprophet/time_net.py:416]: can also skip quantile (for all component functions)?

File: neuralprophet/time_net.py [hunk 16]
```diff
@@ -567,20 +573,12 @@ def forward(self, inputs):
...
-        if self.quantiles is not None:
-            median_quantile = self.quantiles.index(0.5)
-        else:
-            median_quantile = 0
+
         out = (
-            trend
-            + additive_components
-            + trend.detach()[:, median_quantile, :].unsqueeze(dim=1) * multiplicative_components
+            trend + additive_components + trend.detach()[:, 0, :].unsqueeze(dim=1) * multiplicative_components
```
Comment (ourownstory) [inline neuralprophet/time_net.py:574]: add explicative comment

**Review by ourownstory — COMMENTED** [review]
Great work!! Near completion :)

**Comment by ourownstory** [comment]
will be merged in #669
